### PR TITLE
Replace coalesce function with a special form

### PR DESCRIPTION
**Pull request by mbasmanova**
Coalesce needs to be a special form expression and cannot be a function because
it is not supposed to evaluate remaining inputs once there are no nulls. For
example, if column a has no nulls, the following expression should not try to
evaluate `b / 0` and should not fail.

```
coalesce(a, b / 0)
```

This change introduces CoalesceExpr special form and deletes coalesce function.

Also, fix a crash due to null pointer dereference in FlatVector::mutableRawValues.

### CODE REVIEW

**Review by amitkdutta — APPROVED** [review]
Looks good.

File: velox/expression/CoalesceExpr.cpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+    const uint64_t* rawNulls = (*result)->flatRawNulls(*activeRows);
+    if (!rawNulls) {
+      // No nulls left.
+      return;
+    }
+
+    activeRows->deselectNonNulls(rawNulls, 0, activeRows->end());
+    if (!activeRows->hasSelections()) {
+      // No nulls left.
```
Comment (majetideepak) [inline velox/expression/CoalesceExpr.cpp:61]: nit: "no rows left" is probably clear?

**Review by majetideepak — COMMENTED** [review]
@mbasmanova Changes look good to me.

File: velox/vector/FlatVector.h [hunk 2]
```diff
@@ -202,7 +202,7 @@ class FlatVector final : public SimpleVector<T> {
   T* mutableRawValues() {
     if (!values_ || !values_->unique()) {
       BufferPtr newValues =
-          AlignedBuffer::allocate<T>(BaseVector::length_, values_->pool());
+          AlignedBuffer::allocate<T>(BaseVector::length_, BaseVector::pool());
```
Comment (aditi-pandit) [inline velox/vector/FlatVector.h:205]: QQ : Why is this needed for the current change ?

**Review by aditi-pandit — COMMENTED** [review]
Thanks for this change.

Another minor cleanup comment is that we could either move CoalesceExpr class to ControlExpr or promote SpecialForm to its own file since it now has implementations spread out from ControlExpr. Will do it some point.

File: velox/vector/FlatVector.h (hunk 2 above)
Comment (mbasmanova) [reply velox/vector/FlatVector.h:205]: This is needed to avoid a crash when values_ is a nullptr.
